pkg/daemon/update.go

```diff
@@ -1503,22 +1503,27 @@ func (dn *Daemon) writeUnits(units []ign3types.Unit) error {
 			glog.V(2).Infof("Created symlink unit %q to %s", u.Name, pathDevNull)
 
 			continue
-		}
-
-		if u.Contents != nil && *u.Contents != "" {
+		} else if u.Contents != nil && *u.Contents != "" {
 			glog.Infof("Writing systemd unit %q", u.Name)
 			if _, err := os.Stat("/usr" + fpath); err == nil &&
 				dn.os.IsCoreOSVariant() {
 				if err := createOrigFile("/usr"+fpath, fpath); err != nil {
 					return err
 				}
 			}
-			// write the unit to disk
+			// write the unit to fpath
 			if err := writeFileAtomicallyWithDefaults(fpath, []byte(*u.Contents)); err != nil {
 				return fmt.Errorf("failed to write systemd unit %q: %v", u.Name, err)
 			}
 
 			glog.V(2).Infof("Successfully wrote systemd unit %q: ", u.Name)
+		} else {
+			// otherwise fpath should be empty
+			glog.V(2).Info("Cleaning up systemd unit")
+			if err := os.RemoveAll(fpath); err != nil {
```

Comment:
Should this go in `deleteStaleData`? Wasn't sure, but I think cleaning up the symlinks from a mask needs to happen before preset

Reply:
Yes I think attempted removal (in this case, diffing off a mask of a service) should go into deleteStateData I think.

Units are a bit more complex in this way. Just because mask and content are nil does not mean other subfields can be extrapolated. We should instead be checking (for this specific bug) to see whether a mask was previous defined and now gone.